parlai/core/torch_ranker_agent.py

```diff
@@ -0,0 +1,351 @@
+            self.metrics = SharedTable(self.metrics)
+            self.model.share_memory()
+        shared['metrics'] = self.metrics
+
+        # TODO: Consider moving shared candidates to shared memory?
+        shared['fixed_candidates'] = self.fixed_candidates
+        shared['fixed_candidates_vec'] = self.fixed_candidates_vec
+        return shared
+
+    def update_params(self):
+        """Do optim step and clip gradients if needed."""
+        if self.clip > 0:
+            torch.nn.utils.clip_grad_norm_(self.model.parameters(), self.clip)
+        self.optimizer.step()
+
+    def reset_metrics(self):
+        """Reset metrics."""
+        super().reset_metrics()
+        self.metrics['loss'] = 0.0
+        self.metrics['examples'] = 0
+        self.metrics['rank'] = 0
+
+    def report(self):
+        """Report loss and mean_rank from model's perspective."""
+        m = {}
+        examples = self.metrics['examples']
+        if examples > 0:
+            m['examples'] = examples
+            if self.metrics['loss'] > 0:
+                m['loss'] = self.metrics['loss']
+                m['mean_loss'] = self.metrics['loss'] / examples
+            if self.metrics['rank'] > 0:
+                m['mean_rank'] = self.metrics['rank'] / examples
+        for k, v in m.items():
+            # clean up: rounds to sigfigs and converts tensors to floats
+            m[k] = round_sigfigs(v, 4)
+        return m
+
+    def vectorize(self, *args, **kwargs):
+        """Override options in vectorize from parent."""
+        kwargs['add_start'] = False
+        kwargs['add_end'] = False
+        kwargs['split_lines'] = True
+        return super().vectorize(*args, **kwargs)
+
+    def _get_model_file(self, opt):
+        model_file = None
+
+        # first check load path in case we need to override paths
+        if opt.get('init_model') and os.path.isfile(opt['init_model']):
+            # check first for 'init_model' for loading model from file
+            model_file = opt['init_model']
+
+        if opt.get('model_file') and os.path.isfile(opt['model_file']):
+            # next check for 'model_file', this would override init_model
+            model_file = opt['model_file']
+
+        if model_file is not None:
+            # if we are loading a model, should load its dict too
+            if (os.path.isfile(model_file + '.dict') or
+                    opt['dict_file'] is None):
+                opt['dict_file'] = model_file + '.dict'
+
+        return model_file
+
+    def _set_fixed_candidates(self, shared):
+        if shared:
+            self.fixed_candidates = shared['fixed_candidates']
+            self.fixed_candidates_vec = shared['fixed_candidates_vec']
+        else:
+            if self.opt['fixed_candidates_path']:
+                print(f"Vectorizing fixed candidates set from "
+                    f"{self.opt['fixed_candidates_path']}")
+                with open(self.opt['fixed_candidates_path'], 'r') as f:
```

Comment:
A couple things here: the biggest use case for this is interactive mode, in which case the set of fixed candidates is often *huge*, so
- we might want to only encode the fixed candidates with the model *once* at init time
- we might want to save these encoded candidates to a file for loading, so we don't have to repeat this
- you might run into some memory issues, consider batching here and/or forcing this to happen on CPU
(you can see how this is handled in internal)

Reply:
The way it's currently written, it does encode only once at init time. But it doesn't cache the encodings or handle batching. I'll pull up the heavy-lifting code from internal, since I imagine that class will eventually inherit from this one anyway (ideally).